### DISCUSSION TITLE
Debugging Java code with nvim-jdtls and nvim-dap

### DISCUSSION THREAD

**Discussion by britter** [nix-community/nixvim#2534] [Q&A]
Hi,

I'm trying to set up nvim as an IDE replacement for Java development. I'm using the nvim-jdtls plugin to configure `pkgs.jdt-language-server` and am struggling to get the integration between nvim-jdtls and nvim-dap working.

[nvim-dap's documentation](https://github.com/mfussenegger/nvim-dap/wiki/Java#via-nvim-jdtls) states:

> You **do not** have to define `dap.adapters.java` yourself.

However, when I try to debug an application by calling `:DapContinue` the status line says 

> No configuration found for `java`. You need to add configs to `dap.configurations.java` (See `:h dap-configuration`)

My configuration looks like this:

```nix
# java.nix
{
  lib,
  pkgs,
  ...
}: let
  singleFile = dir: let
    files = builtins.attrNames (builtins.readDir dir);
  in
    if builtins.length files == 1
    then "${dir}/${(builtins.elemAt files 0)}"
    else throw "Directory ${dir} does not contain exactly one file.";
in {
  programs.nixvim.plugins.nvim-jdtls = {
    enable = true;
    cmd = [
      (lib.getExe pkgs.jdt-language-server)
    ];
    initOptions = {
      bundles = [(singleFile "${pkgs.vscode-extensions.vscjava.vscode-java-debug}/share/vscode/extensions/vscjava.vscode-java-debug/server")];
    };
  };
}
```

```nix
# debug.nix
_: {
  programs.nixvim = {
    plugins.dap = {
      enable = true;
      extensions.dap-ui.enable = true;
    };
    keymaps = [
      {
        action.__raw = "require('dap').toggle_breakpoint";
        key = "<leader>dt";
        mode = ["n"];
      }
      {
        action.__raw = "require('dap').continue";
        key = "<leader>dc";
        mode = ["n"];
      }
      {
        action.__raw = "require('dap').step_over";
        key = "<leader>ds";
        mode = ["n"];
      }
      {
        action.__raw = "require('dap').step_into";
        key = "<leader>di";
        mode = ["n"];
      }
    ];
  };
}

```

And then another file that include the two and some other configurations.

**Comment by 573** [comment]
@britter, you might need a `dap.configurations.java` attribute section somewhere, look at [recent typical example](https://github.com/khaneliman/khanelivim/blob/bffdb2fe915045eaad356e12e651d3048b30aa10/modules/nixvim/plugins/dap/lua.nix#L57) for how that is done.
(also [plugins.dap.configurations](https://nix-community.github.io/nixvim/plugins/dap/index.html#pluginsdapconfigurations) and [that test](https://github.com/nix-community/nixvim/blob/6645fea915ecf17cf465d544fd42f0f549c00c14/tests/test-sources/plugins/dap/nvim-dap.nix) ([via](https://github.com/nix-community/nixvim/discussions/921#discussioncomment-8065911)))
